### PR TITLE
WIP - GLES: Download depth values via shader

### PR DESCRIPTION
**Pull request by unknownbrackets**
~~This doesn't work on mobile, but I'm not sure why.  It works on desktop (took some tweaking to get the depth values read accurately.)~~ This works on depth and mobile now, at least NVIDIA and Adreno.  Not actually triggered yet.

Obviously, for depth depal cases, we wouldn't download at all.  But making depth downloadable is very useful to compare hardware tests of behavior, and is inevitably used by some games (probably Silent Hill.)

This is a cleanup of an old branch I was using.  Notably, I looked at [Aras' approach](http://aras-p.info/blog/2009/07/30/encoding-floats-to-rgba-the-final/) - it just didn't match the values correctly on desktop, though I'm sure it's a bit faster.  I am worried about scenarios like Phantasy Star text boxes, and anyway inaccurate values will be trouble for tests.

~~Depth itself is working, so I think everything is okay in depth texture land.  But the shader is just getting 0xc34f as the depth on Adreno, always.... somewhat mysteriously (no matter what value I clear it to.)  NVIDIA (GL or GLES) gets it right.~~

~~Any ideas what I'm missing?~~

-[Unknown]

### CODE REVIEW

**Comment by hrydgard** [comment]
Could be yet another Adreno hardware bug (or possibly some obscure GLES thing).

I'm working on an in-app GPU functionality tester, which I will use to (in a roundabout way, don't ask) report the stencil/depth/discard bugs. Looks like another candidate for it.

File: GPU/GLES/DepthBufferGLES.cpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+#define texture2D texture
+#define gl_FragColor fragColor0
+out vec4 fragColor0;
+#endif
+varying vec2 v_texcoord0;
+uniform float u_depthScaleFactor;
+uniform sampler2D tex;
+void main() {
+  float depth = texture2D(tex, v_texcoord0).r;
+  float offset = 0.5 * (u_depthScaleFactor - 1.0) * (1.0 / u_depthScaleFactor);
```
Comment (hrydgard) [inline GPU/GLES/DepthBufferGLES.cpp]: here you can just divide by u_depthScaleFactor, computing the inverse within parenthesis doesn't gain us anything

Comment (unknownbrackets) [reply GPU/GLES/DepthBufferGLES.cpp]: Yeah, I think I just copied this from c++ for now, was trying to get it work first before optimizing or simplifying really.  Could probably pass offset as a separate uniform to make sure the shader is simple.

-[Unknown]

File: GPU/GLES/DepthBufferGLES.cpp [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+  // At this point, clamped maps [0, 1] to [0, 65535].
+  float clamped = clamp((depth - offset) * u_depthScaleFactor, 0.0, 1.0);
+
+  vec4 enc = vec4(16777215.0, 16777215.0 / 256.0, 16777215.0 / 65536.0, 16777215.0 / 16777216.0) * clamped;
```
[the remark this answers is not in the document]
Comment (unknownbrackets) [reply GPU/GLES/DepthBufferGLES.cpp]: Hm, I suppose if they all become infinity or something broken, maybe that explains why every depth value turns into the same result.  Will try.

Would be curious to know if this behaves any different on Mali or PowerVR.

-[Unknown]

[the remark this answers is not in the document]
Comment (unknownbrackets) [reply GPU/GLES/DepthBufferGLES.cpp]: Sending these constants as uniforms fixed it.  Probably better anyway...

-[Unknown]

[the remark this answers is not in the document]
Comment (hrydgard) [reply GPU/GLES/DepthBufferGLES.cpp]: One would think the compiler would evaluate these at full precision, but apparently not then... Bizarre.

**Comment by hrydgard** [comment]
Did you really have to add u_depthTo8 for it to work, or was it more of a might as well do it? Anyway, seems fine.

**Pull request merged**